### ISSUE TITLE
Backport request: Fantasy Zone II - The Tears of Opa-Opa (fantzn2x)

### ISSUE THREAD

**Issue by vaguerant** [libretro/mame2003-plus-libretro#1239]
The original Fantasy Zone was released for Sega System 16 and is emulated in MAME 2003 Plus. The sequel, Fantasy Zone II, was originally a Sega Master System exclusive, but in 2008 Sega contracted M2 to port the game to System 16 hardware.

The new Fantasy Zone II runs on a slightly modified System 16 board, with more RAM than the original boards. It was added to MAME in  mid-2012, well beyond MAME 2003/0.78, but I'm hoping the changes needed are small enough that the 2003 version of the driver could have System 16C/Fantasy Zone II support added without the need to backport too much.

Here's the commit which added Fantasy Zone II to MAME.
mamedev/mame@bac858b0a3f73860847e7430ba97b9c8657ddccb
MAME was still using the old non-commercial license at this point, so even if it isn't directly usable, it should be fine to reference its code.

**Comment by arcadez2003** [comment]
Yeah im familiar with this one as i keep a copy of the game in my set decent wee game but annoyingly missing
a continue mode, short answer here is simply this look at the commit and those tables at the top of the driver
which were created so that many games can share the same code instead of like our ole driver having a unique
code hookup for every single game.

The problem here is as i said we dont have those tables in our ole driver so to add this game and get it up and
running we'd need to create a whole new hookup for it directly im not saying it cannot be done but it certainly
wont be a simple job

Regards

**Comment by vaguerant** [comment]
It's a shame to hear that it's not a quick fix. Feel free to close the issue if the request is unrealistic.

**Comment by mahoneyt944** [comment]
Seems to look like wrestwar though fz2.a2 and fz2.b11 are bigger so I think we'd use rom_continue on these.? Probably completely wrong 🙃

```
ROM_START( fantzn2x )
	ROM_REGION( 0xc0000, REGION_CPU1, 0 ) /* 68000 code */
	ROM_LOAD16_BYTE( "fz2.a7", 0x00000, 0x20000, CRC(94c05f0b) SHA1(53da68a919776a46ae96dbc094ff941308d13613) )
	ROM_LOAD16_BYTE( "fz2.a5", 0x00001, 0x20000, CRC(f3526895) SHA1(3197956608138601192f111d3bcc26662a7d6ec1) )
	/* empty 0x40000 - 0x80000 */
	ROM_LOAD16_BYTE( "fz2.a8", 0x80000, 0x20000, CRC(b2ebb209) SHA1(bd40c90a372ab92a869bdd28d12cf52b45ecc80e) )
	ROM_LOAD16_BYTE( "fz2.a6", 0x80001, 0x20000, CRC(6833f546) SHA1(b4503cdb5bdb1322c34b9da3ff4227c740dad707) )

	ROM_REGION( 0x60000, REGION_GFX1, 0 ) /* tiles */
	ROM_LOAD( "fz2.a14", 0x00000, 0x20000, CRC(1c0a4537) SHA1(3abdf51ea81780309bcfaf12c04efdf7cb15a649) )
	ROM_LOAD( "fz2.a15", 0x20000, 0x20000, CRC(2b933344) SHA1(5b53ea8d58cc3d157aec6926db048359984e4276) )
	ROM_LOAD( "fz2.a16", 0x40000, 0x20000, CRC(e63281a1) SHA1(72379c579484c1ef7784a9598d373446ef0a472b) )

	ROM_REGION( 0x300000, REGION_GFX2, 0 ) //* sprites */
	ROM_LOAD16_BYTE( "fz2.b1",  0x000001, 0x10000, CRC(46bba615) SHA1(b291df4a83d7155eb7606f86ed733c24362a4db3) )
	ROM_CONTINUE( 0x040001, 0x10000 )
	ROM_RELOAD  ( 0x020001, 0x10000 )
	ROM_CONTINUE( 0x060001, 0x10000 )
	ROM_LOAD16_BYTE( "fz2.b5",  0x000000, 0x10000, CRC(bebeee5d) SHA1(9e57e62c6b9136667aa90d7d423fc33ac6df4352) )
	ROM_CONTINUE( 0x040000, 0x10000 )
	ROM_RELOAD  ( 0x020000, 0x10000 )
	ROM_CONTINUE( 0x060000, 0x10000 )
	ROM_LOAD16_BYTE( "fz2.b2",  0x080001, 0x10000, CRC(6681a7b6) SHA1(228df38601ba3895e9449921a64850941715b421) )
	ROM_CONTINUE( 0x0c0001, 0x10000 )
	ROM_RELOAD  ( 0x0a0001, 0x10000 )
	ROM_CONTINUE( 0x0e0001, 0x10000 )
	ROM_LOAD16_BYTE( "fz2.b6",  0x080000, 0x10000, CRC(42d3241f) SHA1(c3240e3e1d7d398e74e76ba65adca6b06f0f67a9) )
	ROM_CONTINUE( 0x0c0000, 0x10000 )
	ROM_RELOAD  ( 0x0a0000, 0x10000 )
	ROM_CONTINUE( 0x0e0000, 0x10000 )
	ROM_LOAD16_BYTE( "fz2.b3",  0x100001, 0x10000, CRC(5863926f) SHA1(0e591c4b85e5d572b3311bec2c1f6d2484204db6) )
	ROM_CONTINUE( 0x140001, 0x10000 )
	ROM_RELOAD  ( 0x120001, 0x10000 )
	ROM_CONTINUE( 0x160001, 0x10000 )
	ROM_LOAD16_BYTE( "fz2.b7",  0x100000, 0x10000, CRC(cd830510) SHA1(8a32a1aa43f8af5e86f552f05da40b6e4ba12495) )
	ROM_CONTINUE( 0x140000, 0x10000 )
	ROM_RELOAD  ( 0x120000, 0x10000 )
	ROM_CONTINUE( 0x160000, 0x10000 )
	ROM_LOAD16_BYTE( "fz2.b4",  0x180001, 0x10000, CRC(b98fa5b6) SHA1(c3f8891f81e80321e2ee5cc1f4d93b1867ed1868) )
	ROM_CONTINUE( 0x1c0001, 0x10000 )
	ROM_RELOAD  ( 0x1a0001, 0x10000 )
	ROM_CONTINUE( 0x1e0001, 0x10000 )
	ROM_LOAD16_BYTE( "fz2.b8",  0x180000, 0x10000, CRC(e8248f68) SHA1(7876945d2baf1d7bdb9cc3a23be9f1a1681cede9) )
	ROM_CONTINUE( 0x1c0000, 0x10000 )
	ROM_RELOAD  ( 0x1a0000, 0x10000 )
	ROM_CONTINUE( 0x1e0000, 0x10000 )
	ROM_LOAD16_BYTE( "fz2.a1",  0x200001, 0x10000, CRC(9d2f41f3) SHA1(54f5dc47d854cd26b108695f55263d8b8c29ce0e) )
	ROM_CONTINUE( 0x240001, 0x10000 )
	ROM_RELOAD  ( 0x220001, 0x10000 )
	ROM_CONTINUE( 0x260001, 0x10000 )
	ROM_LOAD16_BYTE( "fz2.b10", 0x200000, 0x10000, CRC(7686ea33) SHA1(812a638f42500b30f80f9a3956c5eb4553cc35d0) )
	ROM_CONTINUE( 0x240000, 0x10000 )
	ROM_RELOAD  ( 0x220000, 0x10000 )
	ROM_CONTINUE( 0x260000, 0x10000 )
	ROM_LOAD16_BYTE( "fz2.a2",  0x280001, 0x10000, CRC(3b4050b7) SHA1(8c7c8051c577a4b2ca54d7e60c100fbd5391551f) )
	ROM_CONTINUE( 0x2a0001, 0x10000 )
	ROM_LOAD16_BYTE( "fz2.b11", 0x280000, 0x10000, CRC(da8a95dc) SHA1(d44e1515008d4ee302f940ce7799fa9a790799e9) )
	ROM_CONTINUE( 0x2a0000, 0x10000 )

	ROM_REGION( 0x50000, REGION_CPU2, 0 ) /* sound CPU */
	ROM_LOAD( "fz2.a10", 0x00000, 0x08000, CRC(92c92924) SHA1(3c98cea8f42c316405b28ae03469c6876de5e806) )
	ROM_LOAD( "fz2.a11", 0x10000, 0x20000, CRC(8c641bb9) SHA1(920da63961d2f3457c80d4c5f6d4f405374bb23a) )
ROM_END
```

**Comment by arcadez2003** [comment]
So you wanna hang it off Wrestle War we'd still need to call the extra ram via a new memory map
the pcb is basically System B with extra ram hence some people call it System C

i tried something similar last year but used Wonder Boy III instead game never booted right enough
but then again maybe i should have hooked up to the grabbling game instead :)

Anyway @mahoneyt944 give it a go and see how it pans out im busy just now with Xena Crisis trying 
to add support for this recent dump to our ole NeoGeo driver but i cant get it to boot as of yet.

**Comment by mahoneyt944** [comment]
I was wondering if we could use the driver init to install the extra read and write handlers?

**Comment by grant2258** [comment]
yes you can install memory and port handlers if you choose too. You just need to figure out what memory handler you need  if its the 68k be  install_mem_read16_handler and  install_mem_write16_handler check the memory maps with the MRA to get the right size could be 8 16 or 32 depending  cpu

**Comment by mahoneyt944** [comment]
That was my experience when I tried before, though I never hooked the extra ram up either. So its definitely hopeful that it's possible. I'll take another look at it.

**Comment by MistyDreams** [comment]
is this the correct file not even booting for me?

**Comment by arcadez2003** [comment]
Still WIP not playable as yet

**Comment by MistyDreams** [comment]
I dont hear any functioning in the background either that's why i was asking if it was the correct file.

**Comment by arcadez2003** [comment]
Sure with some games they'll at least boot errors and all, while with others if even one part of the emulation is incorrect it'll throw
off the rest hence although it boots nothing is happening.

**Comment by mahoneyt944** [comment]
@arcadez2003 have you tried `SYS16_MRA16_WORKINGRAM2, &sys16_workingram2` added separately from the original working ram? I was looking through aburner, and noticed it also used 2 working rams. Maybe we can add another line to the mem maps with workingram2 instead of lumping it as one.

https://github.com/libretro/mame2003-plus-libretro/blob/755e64ed8c7f1fdc5d69e65084a3355c51107937/src/drivers/aburner.c#L758-L765

**Comment by arcadez2003** [comment]
It might be worth a try

**Comment by MistyDreams** [comment]
maps completely wrong above this will get the game booting. I guessed the soundport as well and added the logged input ports. If someone wants to finish the inputs up feel free. 

```
static MEMORY_READ16_START( fantzn2x_readmem )
    { 0x000000, 0x0bffff, MRA16_ROM },
	{ 0x200000, 0x23FFFF, SYS16_MRA16_WORKINGRAM }, // correct for extra ram board.??
	{ 0x3F0000, 0x3FFFFF, SYS16_MRA16_EXTRAM, }, //rom_5704_bank
	{ 0x400000, 0x40FFFF, SYS16_MRA16_TILERAM },
	{ 0x410000, 0x410FFF, SYS16_MRA16_TEXTRAM }, //*
	{ 0x440000, 0x4407FF, SYS16_MRA16_SPRITERAM },
	{ 0x840000, 0x840FFF, SYS16_MRA16_PALETTERAM },
    // none of these are are confirmed what they do just filled the space
	{ 0xc41000, 0xc41001, input_port_4_word_r }, 
	{ 0XC41002, 0xc41003, input_port_3_word_r },
	{ 0xc41006, 0xc41007, input_port_0_word_r }, // player2
	{ 0xc42000, 0xc42001, input_port_1_word_r }, // dip2
	{ 0xc42002, 0xc42003, input_port_3_word_r }, // dip1
	{ 0xFE0006, 0xFE0006, ww_io_service_r },

MEMORY_END

static MEMORY_WRITE16_START( fantzn2x_writemem )
    { 0x000000, 0x0bffff, MWA16_ROM },
	{ 0x200000, 0x23FFFF, SYS16_MWA16_WORKINGRAM },
	{ 0x3F0000, 0x3FFFFF, SYS16_MWA16_EXTRAM, &sys16_extraram}, //rom_5704_bank
	{ 0x400000, 0x40FFFF, SYS16_MWA16_TILERAM, &sys16_tileram },
	{ 0x410000, 0x410FFF, SYS16_MWA16_TEXTRAM, &sys16_textram },
	{ 0x440000, 0x4407FF, SYS16_MWA16_SPRITERAM, &sys16_spriteram },
    { 0xc40000, 0xc40001, sys16_coinctrl_w },
	{ 0x840000, 0x840FFF, SYS16_MWA16_PALETTERAM, &paletteram16 },
	{ 0xFE0006, 0xFE0006, sound_command_w }, // this was best guess need verfied
MEMORY_END
```

**Comment by mahoneyt944** [comment]
Well I'm glad that the rom loading was correct..... Or at least enough to work here. Once we fill out the mem maps the rest of the way this game should be operational 👍 ...oh and we need to confirm the extra ram too

Nice work

**Comment by MistyDreams** [comment]
its functional with sound just needs inputs added all you need to do is all this memmaps to arcadez zip

**Comment by mahoneyt944** [comment]
@MistyDreams hey i committed this already. Sorry if you were also working on this too. I used arcadez inputs, your mem maps updates with the right ports, and my rom loading and it seems to work right.

Might be some tweaks yet though with the dips, newer mame shows `02` as demo sounds but it doesn't seem to function here.

**Comment by MistyDreams** [comment]
see the post in pull request clearly you dont test. I dont know how you run things but its impossible to get anything done here.

**Comment by mahoneyt944** [comment]
I'm playing the game right now? I test everything. You're looking at old commits. Just be cause a branch is live doesn't mean it's final

![fantzn2x-220530-054714](https://user-images.githubusercontent.com/49591133/170966358-df278726-7f0f-4ced-9e6b-b9a796007616.png)

**Comment by mahoneyt944** [comment]
@MistyDreams see where you comment is? And the commits after the fact?
![Screenshot_20220530-055014](https://user-images.githubusercontent.com/49591133/170967079-6263a3a4-1e2e-46d4-8722-3b6daf2eba23.png)

**Comment by MistyDreams** [comment]
@arcadez2003 here is the zip since its messed up system16 service ports differs.
[opaopa.zip](https://github.com/libretro/mame2003-plus-libretro/files/8797387/opaopa.zip)


@mahoneyt944  that was the memory maps when the game didnt work until I fixed it. I feel like waiting my time explaining this but ill leave it

**Comment by mahoneyt944** [comment]
@MistyDreams you're getting way to hostile here, you were just telling me the game wouldn't boot  and that I don't test things when that's not the case at all with the version that was pushed to the master branch.

The branch you commented on was an initial code dump that I later edited with your memory maps updated to use the inputports that arcadez shared. I then tested the game through multiple levels of play, entered the service mode, and tested all the inputs within that menu.

I'm sure there are tweaks to improve things further but the game is working at this stage which is why it was committed. I plan to credit you, arcadez, and myself for this as well in a future commit btw.

Try not to get so worked up, when multiple people are working on the same code at the same time it can get confusing. We can always make corrections in future commits.

**Comment by MistyDreams** [comment]
None of this is about credits of hostility its about impatience and rushing on your part. This is the prototype so any issues check it against mame then take it there.  All the source for backport came from mame not me per say anyway so i dont need any credit. All the non encrypted sets can be mapped easily enough if you know how dont think there is any missing games though.

**Comment by mahoneyt944** [comment]
I'm a pretty easy going guy. I try to be fair here as much as I can. I'm not sure why you feel I was `rushing`. No one has been actively working on this or anything. I threw out an idea to Arcadez back in December to hook this to wrestwar providing a quick rom loading for it. He then plugged it in real quick and noted that the core wasn't crashing but it wasn't working either. I think both of us figured the mem maps were wrong or perhaps missing the extra ram but it was left on the back burner. 

When you corrected the mem maps, you stated:
```
its functional with sound just needs inputs added, all you need to do is add the memmaps to arcadez zip
```

I was under the impression you were just sharing the mem maps and were leaving the input hook up for someone else. So I hooked the inputs up, and did loads of testing and pushed the commit.

Sorry if this seemed rushed, I felt the game was working to an acceptable level. If there's any further corrections you want to make feel free to submit a PR anytime. Also unless you're adamant about not getting credit, I feel the 3 of us should be, not that these are taken over seriously or anything....at least by the devs here.

**Comment by MistyDreams** [comment]
Im not  adamant anything here. What I will say I could see you where wasting your time changing memory maps the game didnt crash because it didnt run. 

The only thing that can cause a crash is if it is running some way or the wrong way. I dont understand how you never picked up on all the unmapped memory reads that why I told you it was a waste of time. Then I just deciphered the map.

**Comment by mahoneyt944** [comment]
personally I never reviewed the mem maps arcadez provided. What I did was add his shared zip into a branch so that I could begin working on it. At that point it was just a copy paste of where arcadez was at last.

First thing I did was verify all the roms loaded successfully, since the romload was something I threw together in one of the posts above pretty quickly.
```
[MAME 2003+] Opening ROM file: fz2.a7
[MAME 2003+] read_rom_data: All done
[MAME 2003+] Opening ROM file: fz2.a5
[MAME 2003+] read_rom_data: All done
[MAME 2003+] Opening ROM file: fz2.a8
[MAME 2003+] read_rom_data: All done
[MAME 2003+] Opening ROM file: fz2.a6
[MAME 2003+] read_rom_data: All done
[MAME 2003+] Opening ROM file: fz2.a14
[MAME 2003+] Opening ROM file: fz2.a15
[MAME 2003+] Opening ROM file: fz2.a16
[MAME 2003+] Opening ROM file: fz2.b1
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] Opening ROM file: fz2.b5
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] Opening ROM file: fz2.b2
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] Opening ROM file: fz2.b6
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] Opening ROM file: fz2.b3
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] Opening ROM file: fz2.b7
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] Opening ROM file: fz2.b4
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] Opening ROM file: fz2.b8
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] Opening ROM file: fz2.a1
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] Opening ROM file: fz2.b10
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] Opening ROM file: fz2.a2
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] Opening ROM file: fz2.b11
[MAME 2003+] read_rom_data: All done
[MAME 2003+] read_rom_data: All done
[MAME 2003+] Opening ROM file: fz2.a10
[MAME 2003+] Opening ROM file: fz2.a11
[MAME 2003+] Succesfully loaded ROMs.
```

At this point you shared the corrected map, so I just updated the branch then hooked up the inputs like you noted as needing finished.

**Comment by mahoneyt944** [comment]
And actually, one of the first things I did after creating that branch was fix an entry in the write map arcadez had... it was entered as a read. https://github.com/libretro/mame2003-plus-libretro/pull/1353/commits/3527b64bc70e07b43cad8aaafede51ca72849ad3

Without actually sorting through the maps contents at this point, I figured it may have just been a simple typo that was the issue, so I added the branch to the buildbot to test and see.

But yeah I didn't really review the map up front since I didn't make it. It was just a starting point of what was shared here.

**Comment by MistyDreams** [comment]
The only problem is your approach when someone fixes something and you add one thing you feel a need to commit it. Instead of just sharing it here and getting feedback or even putting a pull request it and asking for testing. It a lack of development respect/decorum and makes collaborating near impossible. I guess everyone does things a different way but this makes for more work and buggy commits in general. Just my opinion we all have our own.

**Comment by arcadez2003** [comment]
@mahoneyt944 upon reflection we did take the code @MistyDreams created and was working on with regards to
Marble Madness II and then we all piled in basically rather than being patient and waiting to see what they came up
with on their own given the proper space and time to do so.

im sorry if that upset you @MistyDreams but i did have my reasons as i explained earlier help was offered which i felt
had to be taken at the time, and then also there is the matter of this game being drama attracting over and above what
we normally get as per this core at the best of times..

But look if your up for it there is a nemesis of mine a game i've been trying to get playable in this core off and on
for nigh on three years would you like to take a look at it to see what's going wrong.?? i 100% guarantee no one
will butt in and you'll be 100% left alone to work on this should you choose to.

**Comment by arcadez2003** [comment]
Anyway with regards to the extra ram im not really sure what it does and with my ole segasb.c driver i dont even have it
totally hooked up entirely as i was missing some code whicj could not be backported however the game seems to function
correctly.

I've heard that the game can fail a check on the later levels and crash maybe that's due to extra ram.?? i dunno but i'll
give this game a full on playtest shortly but the lack of a continue mode can make getting to last few levels kinda tricky
as this is not a easy game.

But again im impressed ya's got it to work in the first place

**Comment by mahoneyt944** [comment]
@arcadez2003 That's a fair enough assessment from the view point of a contributor....and I can see that too.

Though like you, I know the backlash and the behind the scenes stuff that goes on.... And honestly in this particular case I was under the impression Mistydreams was just sharing the mem maps for one of us to finish up. So that's what I did. I'm basing this assumption on his comment when he shared the corrected mem maps
```
If someone wants to finish the inputs up feel free.
```

**Comment by mahoneyt944** [comment]
Yeah just restart the core once you change the dip and it should be what you need to survive ha. I may be able to write an invisibility code for this game too if need be.

**Comment by mahoneyt944** [comment]
I may have found a bug, looks like we are missing some of these upgrades. 

![fantzn2x-220530-095328](https://user-images.githubusercontent.com/49591133/171006811-725b3d38-6c1a-45dc-8ac4-828aa7b0f6dd.png)

Ours shows like this
![fantzn2x-220530-095954](https://user-images.githubusercontent.com/49591133/171007928-68604010-6294-41af-8567-99ff015b4d01.png)

And if you click on where it should be it says sold out
![fantzn2x-220530-100012](https://user-images.githubusercontent.com/49591133/171008014-17eaadb1-29cd-41c9-8d35-c6fd5bfde8c5.png)

**Comment by mahoneyt944** [comment]
maybe the spriteram?

**Comment by mahoneyt944** [comment]
Yes that works right
![fantzn2x-220530-100955](https://user-images.githubusercontent.com/49591133/171009808-961df3f1-de28-4e01-b1c8-fc6a6c76bc98.png)

**Comment by arcadez2003** [comment]
Lovely

**Comment by mahoneyt944** [comment]
This caught my eye.
https://github.com/libretro/mame2003-plus-libretro/blob/df36e0e2cf517b057d210a3de8c875368e226766/src/drivers/system16.c#L21

Maybe this needs it too?

**Comment by arcadez2003** [comment]
Ah bummer but i think that sprite address is still correct however as per the other games no matter the values they
always end with three F's

**Comment by mahoneyt944** [comment]
@arcadez2003 
Yeah probably but I'll wait to see what @MistyDreams thinks since it's his memory map.

**Comment by mahoneyt944** [comment]
Like wise this is probably 
```
{ 0xfe0006, 0xfe0007, ww_io_service_r },
```

**Comment by arcadez2003** [comment]
Yeah just giving my opinon is all :)

**Comment by arcadez2003** [comment]
Let's hope the issue is confined to the driver as if we have to start looking at the different sprite calls this could become
a nightmare Wrestle War uses Shinobi sprites there are other routines though for some of the different games they can
be found in vidhrdw/sys16spr.c

**Comment by mahoneyt944** [comment]
Yeah hopefully not, what do you think about `ww_io_service_r` in the reads. This function returns inputport2 | workingram. Seems odd to me. All the other io's use high values not at the sound address? (0xfe0006, 0xfe0007)

**Comment by MistyDreams** [comment]
@arcadez2003 this should work without the wrestlewar init.  None of the ports are confirmed (apart from input ) and banking needs looked at.  I did mention all the ports need tested and checked

**Comment by mahoneyt944** [comment]
@MistyDreams we refer to "your maps" as the maps you shared with the group here. It's just a reference to which maps, not ownership.

**Comment by mahoneyt944** [comment]
@MistyDreams I'm a little confused as to what this read is doing. It references the service port (port 2)
```
{ 0xfe0006, 0xfe0007, ww_io_service_r }
```

But we are reading the service port here and this is confirmed to work without the above read 
```
{ 0xc41000, 0xc41001, input_port_2_word_r }, // service
```

**Comment by arcadez2003** [comment]
> @arcadez2003 this should work without the wrestlewar init. None of the ports are confirmed (apart from input ) and banking needs looked at. I did mention all the ports need tested and checked as well as the video code. It not my maps its mame maps. The usual suspects mirroring and input ports and incompatible video code!

If your just posting up WIP and want to be left alone to work on this in your own time then that's all you have to say
afterall too many cooks spoil the broth :) i was always solo when it came to my xbox projects apart from asking the
FBN boys for the odd bit advice here and there.

**Comment by MistyDreams** [comment]
I dont mind just dont commit to master till its done anyone can do anything in a branch no harm done else things get confusing its already done so knock yourselves out. I need to get up to speed on this driver anyhow its day and night compared to what im using atm

**Comment by MistyDreams** [comment]
there is some ram not being read in the upper section will youll need to look into that as well

**Comment by MistyDreams** [comment]
@mahoneyt944 I think your taking the meaning of input ports meaning for controls only

**Comment by arcadez2003** [comment]
I know what ya mean in my xbox project i split the Sega code into two parts a best of both worlds kinda deal i kept
the games that were fine grahically in system16.c and i created a new segas16b.c driver to support games that were
either broken or were not supported in MAME84 as that is the MAME version we use..

So technically i have two Sega cores the one we use and another from MAME94 i run them side by side but alas we
cannot do that here due to code inbalances, so yeah it's better to drop games into the drivers they were created for
rather than attempt to shoehorn them into an older version.

But sometimes that is where the fun is ;)

**Comment by MistyDreams** [comment]
Very true will look some  avenues on my branch in time will tie in with you when i get some results. Need to look into sound latches ect as well codes full of workarounds confusing but fun

**Comment by mahoneyt944** [comment]
> @mahoneyt944 I think your taking the meaning of input ports meaning for controls only

I'm not sure what you're replying to here. My comments about `{ 0xfe0006, 0xfe0007, ww_io_service_r }`?  I was trying to see what this is actually doing since I can comment it out without any noticable differences while testing. So if it's actually doing something, I was curious as to what? Just making sure this line is doing what it's intended to.

**Comment by MistyDreams** [comment]
@mahoneyt944  you can drop it isint needed

**Comment by arcadez2003** [comment]
Ah i was starting to wonder if it was the tile layer that handled that rather than the sprites well it's good your onto the issue

**Comment by MistyDreams** [comment]
@arcadez2003  
	{ 0xc40000, 0xc40001, sys16_coinctrl_w }, 

need to find out what this is out putting and fix it per what happens here to our equiv think thats our issue after looking about

https://github.com/mamedev/mame/blob/9cfc798a67baa9f4344e55dcc7da8d139f286406/src/mame/drivers/segas16b.cpp#L1099-L1127

**Comment by arcadez2003** [comment]
Sure i can look into it no problem but maybe not until the weekend got a lot on this week

**Comment by MistyDreams** [comment]
I did add some notes on my findings let me get the link. I need to make notes or i forget what ive done lol 

https://github.com/MistyDreams/mame2003-plus-libretro/blob/25fcb89093ced9c7b631006972bf9efaabcf3355/src/drivers/system16.c#L6837-L6855

**Comment by mahoneyt944** [comment]
@MistyDreams @arcadez2003 I was playing with the inits last night. Probably need to look through what we actually need. This seemed to be more or less the same as what we get using wrestwar inits.
```
static MACHINE_INIT( fantzn2x )
{
	sys16_update_proc = wrestwar_update_proc;
	sys16_wwfix = 1;
}

static DRIVER_INIT( fantzn2x )
{
	machine_init_sys16_onetime();
}
```

**Comment by MistyDreams** [comment]
> Could probably make a `fantzn2_update_proc` with the correct banking too

where are you going to get this information ? how did you get it for wrestle wars?

**Comment by MistyDreams** [comment]
@arcadez2003 its fine refreshable is updated in the coin read

**Comment by mahoneyt944** [comment]
I was just testing different inits out and then started stripping back wrestwar to see what changes occured. I tested out some different proc's from our driver and found that most use a similar formula but with slight tweaks for each game. `set_tile_bank( sys16_extraram[1] )` is important for our code to work, but as arcadez mentioned might need modified some. `sys16_extraram[1]` seems to be the title screen sprites that are bugged out. `sys16_extraram[0]` seems to be the background in game

**Comment by MistyDreams** [comment]
@mahoneyt944  we know what wrestle wars is doing you need to keep in mind ww is system16b_i8751 and fz2 is system16c can try blind that does work sometimes

**Comment by arcadez2003** [comment]
> @arcadez2003 its fine refreshable is updated in the coin read the notes above say where the problem happens

Yeah i was looking at the link you sent last night im my ole segasb it would be 			

if (!disable_screen_blanking)
				system16b_set_draw_enable(data & 0x20);

here as you said

		sys16_refreshenable = coinctrl & 0x20;

Since i didn't have much time today i thought i'd look at the tiles :) anyway seems like your on top of things.

**Comment by mahoneyt944** [comment]
I've been testing this
https://github.com/libretro/mame2003-plus-libretro/compare/T2

Seems like everything is working except the missing sprites on the upgrade sprites. Though I haven't looked for overflows.

**Comment by MistyDreams** [comment]
seesh your patient.

did you go through all the options here?

```
sys16_textram:
type1		type0			function
---------------------------------------
0x74f		0x740			sys16_fg_page
0x74e		0x741			sys16_bg_page
			0x742			sys16_fg2_page
			0x743			sys16_bg2_page

0x792		0x748			sys16_fg_scrolly
0x793		0x749			sys16_bg_scrolly
			0x74a			sys16_fg2_scrolly
			0x74b			sys16_bg2_scrolly

0x7fc		0x74c			sys16_fg_scrollx
0x7fd		0x74d			sys16_bg_scrollx
			0x74e			sys16_fg2_scrollx
			0x74f			sys16_bg2_scrollx

			0x7c0..0x7df	sys18_splittab_fg_x
			0x7e0..0x7ff	sys18_splittab_bg_x
````

**Comment by mahoneyt944** [comment]
No I did not test other options yet.

In regards to your demeanor, you asked that other developers not push to master and work on things in separate branches until a group consensus was made. I shared my branch with the group, which is working better than the current master is, though with some issues to sort still.

I'm going to remind you that LibRetro will not stand for inappropriate behavior towards other developers.

**Comment by MistyDreams** [comment]
> I'm going to remind you that LibRetro will not stand for inappropriate behavior towards other developers.

what ?

**Comment by m4xw** [comment]
Now kiss

**Comment by mahoneyt944** [comment]
Lol were just gonna be a couple of Fonzies around here 😂

**Comment by MistyDreams** [comment]
@arcadez2003  ill update you on the megaplay thing with an @ from my github if i find anything.

@mahoneyt944 

```
set_fg_page( sys16_textram[0x740] );
set_bg_page( sys16_textram[0x741] );
//set_fg_page1( sys16_textram[0x742] );
//set_bg_page1( sys16_textram[0x743] );

sys16_fg_scrolly= sys16_textram[0x748];
sys16_bg_scrolly= sys16_textram[0x749];
sys16_fg2_scrolly= sys16_textram[0x74a];
sys16_bg2_scrolly= sys16_textram[0x74b];

sys16_fg_scrollx= sys16_textram[0x74c];
sys16_bg_scrollx= sys16_textram[0x74d];
sys16_fg2_scrollx= sys16_textram[0x74e];
sys16_bg2_scrollx= sys16_textram[0x74f];

sys18_splittab_fg_x= &sys16_textram[0x7c0];
sys18_splittab_bg_x= &sys16_textram[0x7e0];

//set_tile_bank( sys16_extraram[1] );
sys16_tile_bank0 = sys16_extraram[0]&0xf;
sys16_tile_bank1 = sys16_extraram[1]&0xf;
```
will get your closer. 

@m4xw not sure what ive actually done for that comment but ill refrain from participating here have enough drama in my life with the wife lol.

**Comment by mahoneyt944** [comment]
Ok clean slate here,

The T2 branch isolates the machine and driver init as well as creates a fantzn2x_update_proc that has modified tile banking. This seems to fix all the graphic issues that were happening on the demo screens but does not fix the missing sprites on the parts upgrade screen.

I think isolating this game in this manor is a good idea.

**Comment by mahoneyt944** [comment]
> ```
> set_fg_page( sys16_textram[0x740] );
> set_bg_page( sys16_textram[0x741] );
> //set_fg_page1( sys16_textram[0x742] );
> //set_bg_page1( sys16_textram[0x743] );
> 
> sys16_fg_scrolly= sys16_textram[0x748];
> sys16_bg_scrolly= sys16_textram[0x749];
> sys16_fg2_scrolly= sys16_textram[0x74a];
> sys16_bg2_scrolly= sys16_textram[0x74b];
> 
> sys16_fg_scrollx= sys16_textram[0x74c];
> sys16_bg_scrollx= sys16_textram[0x74d];
> sys16_fg2_scrollx= sys16_textram[0x74e];
> sys16_bg2_scrollx= sys16_textram[0x74f];
> 
> sys18_splittab_fg_x= &sys16_textram[0x7c0];
> sys18_splittab_bg_x= &sys16_textram[0x7e0];
> 
> //set_tile_bank( sys16_extraram[1] );
> sys16_tile_bank0 = sys16_extraram[0]&0xf;
> sys16_tile_bank1 = sys16_extraram[1]&0xf;
> ```
@MistyDreams  this seems to do weird things to the image
![fantzn2x-220531-215735](https://user-images.githubusercontent.com/49591133/171312978-d4f9ad08-e9a3-4f9a-acaa-7cfb40288937.png)

**Comment by mahoneyt944** [comment]
I think I see what's going on here. On the left side of the screen the sprites get cut off before going off the screen, so there's a dead zone. The sprites should travel to the last pixel of the sprite not just disappear.

See FBN for reference on the red guys
![fantzn2x-220601-011441](https://user-images.githubusercontent.com/49591133/171333855-a14d7204-3473-4dac-bb9d-fe92c9e5f7dd.png)

Our red guy sprite splits just before it disappears
![fantzn2x-220601-013044](https://user-images.githubusercontent.com/49591133/171334480-9aa9d7ff-16f5-4e67-a82b-50e922a5fbcb.png)

This seems as if we need to adjust the memory range possibly? Or add a missing one?

**Comment by mahoneyt944** [comment]
I don't believe it's the sprxoffset, tested this already, our sprite offset is in sync with FBN so I feel pretty good about that. Though I guess if it's combined with other configurations then maybe.

Also tried altering the fgxoffset without any luck, but you may want to experiment with this still.

I did not test sys16_spritelist_end. I'm not sure what this is tbh, though I think the sprites are fine just getting cut off too early. If we had a sprite that was missing features id think otherwise.

I also believe the mem maps addresses to be correct as well.

So I'm out of ideas but I feel this is probably one of those one liner fixes. Just can't seem to get it. :(

**Comment by mahoneyt944** [comment]
Maybe a fg priority and/or priority mode?

**Comment by MistyDreams** [comment]
should we be using the system18 video though? We are using the system16 one atm looking at the code video code

**Comment by mahoneyt944** [comment]
Ok I grabbed the system18 bits into the fant branch here to test. It seems to behave in the same manor though with the deadzone present. Is this the case for your branch as well?

**Comment by MistyDreams** [comment]
well if you want to update the video code your your in for a lot of testing seems someones been hard at work to  getting things working in the past. Unless you want to update the code mark it as imperfect gfx

**Comment by MistyDreams** [comment]
@arcadez2003 is assume you came up with that fix for wrestle fest ?

**Comment by arcadez2003** [comment]
> @arcadez2003 is assume you came up with that fix for wrestle fest ?

Nope i was at the fitba Scotland mince as usual

**Comment by MistyDreams** [comment]
@arcadez2003 i mean in the past I see a fix was put in for that game specifically. Im not too interested in taking this on either if you remove the hacks youll need to update the drivers

**Comment by mahoneyt944** [comment]
This really feels like a parameter issue when drawing. The top, bottom, and right sides all allow sprites to roll off screen as they should. The left side just has a small strip where sprites can not be drawn. Seems like we need to extend this parameter on the left side of the screen to allow drawing here. Maybe we can adjust this in the video code with an offset or something? @dinkc64, I'm guessing arcadez inquired about this issue with you already, if not I'm sorry to have bothered you 🙃, but I'm curious about your opinion. One of the assumptions was sprite wrapping, but if this were the case wouldn't we see sprites warping onto other sprites or something? This issue is very consistent, as if the sprites hit a wall where the video code thinks its no longer required to draw.

**Comment by mahoneyt944** [comment]
One of the odd things is that the service menu script test allows all sides to work correctly with the sprite
![fantzn2x-220602-014704](https://user-images.githubusercontent.com/49591133/171561551-8f778710-d416-4d95-b981-b59ea8a918af.png)

**Comment by arcadez2003** [comment]
> @arcadez2003 i mean in the past I see a fix was put in for that game specifically. Im not too interested in taking this on either if you remove the hacks youll need to update the drivers. I tale it we have all sysem16 on some form ie bootlegs now ?

It wasn't just Wrestle War quite a few Sega games benefited from it, i cant mind all of them off the top of my head but
removing this update will make no difference as my MAME72 test core doesn't have it and still Fantasy Zone II has that
sprite bug.

**Comment by mahoneyt944** [comment]
What I don't understand is why 3 sides of the screen are working but the left side has a cutoff. There has to be a shift or offset or something to account for this. We just can't drawn there, while in game anyways.

**Comment by MistyDreams** [comment]
try setting sys16_18_mode = 1

**Comment by mahoneyt944** [comment]
Sure, 1sec.

**Comment by mahoneyt944** [comment]
That's the default anyhow for system18, so should be active already.
https://github.com/libretro/mame2003-plus-libretro/blob/755e64ed8c7f1fdc5d69e65084a3355c51107937/src/vidhrdw/system16_vidhrdw.c#L1016

**Comment by MistyDreams** [comment]
have you tried setting it with system16 video ?

**Comment by arcadez2003** [comment]
Video update system 16 sys16_sprxoffset = -0xb8; by default, this is obviously not correct for this game hence the sprites
are cut off on the far left panel, this is my two cents but i could be wrong :)

**Comment by MistyDreams** [comment]
I hope your right might luck it out and the bootleg will work. Im not willing spend any more time fixing one game if it was a bigger picture maybe.

**Comment by arcadez2003** [comment]
Well im thinking with the correct sprite offset the game will be fine it's finding it though eh, as things stand we need to
alter this slightly sys16_sprxoffset = -0xb8 to shift the sprite layer slightly to the left.

**Comment by arcadez2003** [comment]
I did put -15 to shift the sprites across to the left more just to see if we'd still get that cutoff and it still happens well im outta ideas and dont fancy trying to rewrite some sprite code most likely to fit in with this game as it's not worth it really.

@mahoneyt944 just add an imperfect gfx flag to this one and let's be done with it imo

P.S humour me here these is a fix for Wrestle War it's sprite list end related i think as a last roll of the dice try 
turning it off to see if it's sorts the problem //sys16_wwfix = 1;

**Comment by mahoneyt944** [comment]
> can easily add key presses to increase/decrese the value

Yeah might be worth a try, should be able to add key presses to the video update and use the usrintf_showmessage to display the value

**Comment by MistyDreams** [comment]
sorted on my branch one line fix

**Comment by mahoneyt944** [comment]
@MistyDreams nice work 👍. Did you get a chance to test any other games using sys16_sprite_shinobi? Hopefully this won't break any other games since sys16_sprite_shinobi is the default for system16. If it does we can probably initialize this for just this game.

Maybe by chance, this will fix bugs in other games. But who knows.

**Comment by mahoneyt944** [comment]
@MistyDreams testing all these games doesn't sound fun, maybe we should hook this up separately? Created a PR #1359 to do just that, unless you feel confident that all these games will work with that change.

**Comment by mahoneyt944** [comment]
Also after some clean up in the driver, I wonder if we should be using this for wrestwar too. Hwchamp and fantzn2x share this. Maybe wrestwar as well?
```
sys16_tile_bank0 = sys16_extraram[0]&0xf;
sys16_tile_bank1 = sys16_extraram[1]&0xf
```

**Comment by MistyDreams** [comment]
pretty sure this is good for the sprites all round however i looked beyond this game

**Comment by mahoneyt944** [comment]
Yeah that's my concern, probably best to just separate it for now so we don't have test everything. But I'm also trying to clean up the driver a little too.

I'm going to try that fix with some other games to see if it helps or not.

**Comment by arcadez2003** [comment]
@MistyDreams lovely work bigman i was a little downhearted last night but was planning to look at the sprite code
today but it now seems i wont have to :) far be it for me to interfere but for the record all i did here was paste in
the new sprites for fantzn2x hook it via system16.h and then like so in the driver......

static void fantzn2x_update_proc( void ){
	set_fg_page( sys16_textram[0x740] );
	set_bg_page( sys16_textram[0x741] );
	sys16_fg_scrolly = sys16_textram[0x748];
	sys16_bg_scrolly = sys16_textram[0x749];
	sys16_fg_scrollx = sys16_textram[0x74c];
	sys16_bg_scrollx = sys16_textram[0x74d];

	sys16_tile_bank1 = sys16_extraram[1]&0xf;
	sys16_tile_bank0 = sys16_extraram[0]&0xf;

}

static MACHINE_INIT( fantzn2x ){
	sys16_spritesystem = sys16_sprite_fantzn2x;
	sys16_bg_priority_mode=2;
	sys16_bg_priority_value=0x0a00;
	sys16_update_proc = fantzn2x_update_proc;
	sys16_wwfix = 1; //*
}

static DRIVER_INIT( fantzn2x ){
	machine_init_sys16_onetime();
	sys16_bg1_trans=1;
	sys16_MaxShadowColors=16;
	sys18_splittab_bg_y=&sys16_textram[0x0f40];
	sys18_splittab_fg_y=&sys16_textram[0x0f00];
	sys16_rowscroll_scroll=0x8000;
}

Seems the simpilest way TBH although as @mahoneyt944 said some other games might just benefit from using the
new sprite code

**Comment by mahoneyt944** [comment]
@vaguerant this game is working now if you want to try it out. Closing this issue.

I'm going to start a new one for wrestwar though

**Comment by arcadez2003** [comment]
Nice job on this guys someone posted on retropie that it playes great for them as it was too slow in FBN and later MAME
i love posts like that it just goes to show our faster core and in this case faster Sega code ended up being a real boon
for that user.

And that's one of the reasons i like working on this core, they did mention however that the high scores dont save
i suppose if none of the other Sega 16 games save either then it's just the lack of save support in the driver but
if they all do.?? then we might have missed something.?? one to check on a rainy day.

**Comment by MistyDreams** [comment]
retropie name makes me want to eat something! haha

**Comment by mahoneyt944** [comment]
I read somewhere that a lot of games use nvram that's not hooked up, but you could scout newer hiscore files for entries

**Comment by arcadez2003** [comment]
According to the user Alien Syndrome and the original Fantasy Zone save the highscores so maybe we missed something.??

**Comment by MistyDreams** [comment]
think the workram will be the eprom to be honest

**Comment by mahoneyt944** [comment]
Ahh good reference, fantzone is in the hiscore.dat
https://github.com/libretro/mame2003-plus-libretro/blob/7ff51a1524b01da721f7768b67ecc3a8e69b55a8/metadata/hiscore.dat#L11523-L11526
So we'll probably need to add the new games.

**Comment by MistyDreams** [comment]
you need a nvhandler for workram have different sizes now
